<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Asynchronous JavaScript</title>
</head>
<body>
    <h1>Asynchronous JavaScript</h1>
    <script>

        ////////////////////////////
        // A callback hell
        // function getRecipe() {
        //     setTimeout(() => {
        //        const recipeID = [523,883,432,974];
        //        console.log(recipeID);
                
        //         setTimeout((id) => {
        //         console.log(`my id is ${id}`);
        //          const recipe = {
        //              publisher: 'lukas'
        //          }   
        //             setTimeout( publisher => {
        //                 const recipe2 = {
        //                     title: 'Italian Pizza',
        //                     publisher: 'Jonas'
        //                 }
        //                 console.log(`Our title is ${recipe2.title} and publisher ${publisher}`);


        //             },1000, recipe.publisher);

        //         },1000, recipeID[2]); 

        //     },1500);
        // }

        // getRecipe();


            //we either call resolve func is success or reject func, setTimeout will always be success. , so we can just call resolve
      const getIDs = new Promise((resolve, reject) => { // we pass a func called executer, it will be immediate called once the promise is created
            setTimeout( () => {
                resolve([523,883,432,974]);
                // after timer finishes we say that promise is succ. and that we want to return the data
            },1500);
        });


        const getRecipe = recID => { // we did this one as a func coz we needed to pass a recID
           return new Promise( (resolve, reject) => {
                setTimeout(ID => { 
                    const recipe = {
                        title: 'Fresh tomato pasta',
                        publisher: 'lukas'
                    };
                    resolve(`${ID}: ${recipe.title}`); // **

                },1500,recID);  
           }); 
        };

        const getRelated = publisher => {
            return new Promise((resolve, reject) => {
                setTimeout(pub => {
                    const recipe = {
                        title: 'Italan pizza',
                        publisher: 'Jonas'};
                        resolve(`${pub}: ${recipe.title}`);
                },1500, publisher);
            })
        };

        // we produce our promise which is stored into getIDs var, and now we consume our promise, we can use 2 methods for it
        // then and the catch methods,all of the promise object inherits these methods
    
        //if promise is successfull , on our promise object we use then method,which is to handle fulfilled prome, then we can handle that
          
          //we consume our promises with the chain of thens
            // getIDs.then(IDs => { // data from our resolve func 
            //     console.log(IDs);
            //     //our next promise
            //     return getRecipe(IDs[2]);// this will return promise and next then will handle the result of that promise
            // })
            // .then(recip => {
            //   console.log(recip);  // ** is returned here
            //   return getRelated('lukas'); // a new promise
            // })
            // .then(recipe => {
            //     console.log(recipe);
            // })
            // .catch(error => {
            //     console.log(error);
            // });


            // async function getRecipesAW() {
            //     //await will stop the code from executing, at this point until the promise is fulfilled
            //     //we wait for this promise here to resolve and assign the value to IDs var
            //     const IDs = await getIDs; 
            //     console.log(IDs);
            //     //our second promise
            //    const recepi = await getRecipe(IDs[2]);
            //     console.log(recepi);
            //     //third Promise
            //     const related = await getRelated('Lukas');
            //     console.log(related);
            //     return recepi;
            // }
            // // async  func automatic. returns a promise with resolved value of recepi and we can use then in to which we pass a callback func where the argument is the resolved value of the promise
            // getRecipesAW().then(result => console.log(`${result}`));

            
            // function getWeather(woeid){ 
            //     //fetch gets our data and returns a promise
            //     fetch(`https://cors-anywhere.herokuapp.com/https://www.metaweather.com/api/location/${woeid}/`)
            //     .then(result => { //fetch AJAX request will be called result here in this callback func
            //         console.log(result); // we convert the body from json to JavaScript 
            //         return result.json(); // returns a promise coz it happens Asyn in the background, it can take some time
            //     })
            //     .then(data => {
            //         //console.log(data);
            //         const today = data.consolidated_weather[0];
            //         console.log(`Temperatures in ${data.title} stay between ${today.min_temp} and ${today.max_temp}`);
            //     })
            //     .catch(error => console.log(error));  
            // }
            // getWeather(2487956);
            // getWeather(44418);


            async function getWeatherAW(woeid){

                try{ 
                    //code stops here with the await expression coz entire func is running  int he background
                    const result = await fetch(`https://cors-anywhere.herokuapp.com/https://www.metaweather.com/api/location/${woeid}/`);
                    const data = await result.json();
                    const tomorrow = data.consolidated_weather[1];
                    console.log(`Temperatures tomorrow in ${data.title} stay between  ${tomorrow.min_temp} and ${tomorrow.max_temp} `);
                    return data;
                } catch(error) {
                        alert(error);
                    }
            }
            //getWeatherAW(2487956);
            
            let dataSA; 
             getWeatherAW(2487956).then( data => {
                dataSA = data;
                console.log(data);
             });

             //!!!!!can not do it like this
             //let dataSA = getWeatherAW(2487956);
             //console.log(dataSA);
            
    </script>
</body>
</html>